### ISSUE TITLE
Averaging integers gives a float

### ISSUE THREAD

**Issue by carstenbauer** [ffreyer/BinningAnalysis.jl#2]
Try creating a binner with `zero(Int64)` and add more than one element. This will fail because the averaging of the first compressor, `0.5 * (C.value + value)`, will produce a `Float64` and there is no method `push!` for `Float64` for this binner.

**Comment by carstenbauer** [comment]
Maybe check in `push!` whether one is pushing to the first level. If yes, only allow values of type `T`, otherwise allow any type. We should be the only ones pushing to higher levels anyway, so probably we should rename `push!` (with a level argument) to `_push!` to avoid exporting it.

**Comment by carstenbauer** [comment]
However, this won't solve it fully, as the problem propagates: from level 2 on, `x_sum` and `x2_sum` will get `Float64`s which they can't handle as they have `eltype(x_sum) == Int64`.

What I did in MonteCarloObservable.jl is to apply a heuristic: If I get a type `<: Real` then make `x_sum` etc. take `Float64` and if I get `<: Complex` then make `x_sum` store `ComplexF64` values. Although not strictly valid it should handle most practical cases. To cover the remaining cases, I gave the user an option to specify the `meantype` explicitly.

**Comment by ffreyer** [comment]
I agree that the recursive push! shouldn't be exported. There's no need to use it.

Promoting `x_sum` to `Float64` or maybe just the whole `LogBinner` sounds like good solution. I'm leaning towards the latter as it keeps the types consistent.

**Comment by carstenbauer** [comment]
> Promoting x_sum to Float64 or maybe just the whole LogBinner sounds like good solution. I'm leaning towards the latter as it keeps the types consistent.

That's what I did in #3.

After #3 is merged, what is left to do here is to give the user the option to set the type parameter explicitly (avoid the heuristic).
